### ISSUE TITLE
Restore timepoints and source group when using the editor mode

### ISSUE THREAD

**Issue by NicoKiaru** [bigdataviewer/bigdataviewer-playground#272]

**Comment by NicoKiaru** [comment]
This is in fact fixed in https://github.com/BIOP/bigdataviewer-selector/commit/44321479fc44c35348937b6dae849dfd0bca593b

**Comment by NicoKiaru** [comment]
@tpietzsch just FYI, when BdvOverlay#removeFromBdv is called in a bdv window, there's a lot of things from the bdv state which are changed: the number of timepoints is reset to one, the sources within the groups are reset, and probably some other things.

To go around this issue, I have to save the state snapshot and restore "manually" parts of the state. This is okayish, but it would be better if I could avoid that (see https://github.com/BIOP/bigdataviewer-selector/blob/44321479fc44c35348937b6dae849dfd0bca593b/src/main/java/ch/epfl/biop/bdv/select/SourceSelectorBehaviour.java#L196-L205)

Not that important, but I wanted to mention it to you since it impairs the user experience.

**Comment by NicoKiaru** [comment]
TODO: do a clean issue in bigdataviewer-vistools

**Comment by NicoKiaru** [comment]
Todo fix the SourceGroup workaround in bigdataviewer-selector once https://github.com/bigdataviewer/bigdataviewer-core/issues/177 is fixed

**Comment by tpietzsch** [comment]
@NicoKiaru I merged your fix for https://github.com/bigdataviewer/bigdataviewer-core/issues/177

Can you explain how you use timepoints?

When BdvOverlay#removeFromBdv is called, vistools recomputes the numTimepoints as the max `getNumTimepoints()` over all sources (that are left) https://github.com/bigdataviewer/bigdataviewer-core/blob/783b7aee870348f231fdf8c7b214a5fa8a31e9df/src/main/java/bdv/util/BdvHandle.java#L291-L298

So it looks like, either, your overlay had more timepoints than the other sources, or you set the numTimepoints directly in the viewerstate (bypassing vistools BdvSources)?

**Comment by NicoKiaru** [comment]
Most of the time, I build directly the `SourceAndConverter` object, which, I think, has no convenient way to return a number of timepoints ? 

And then each source is added via:

https://github.com/bigdataviewer/bigdataviewer-playground/blob/ee30c16ff9073f4741b8138777385027ea123097/src/main/java/sc/fiji/bdvpg/scijava/services/SourceAndConverterBdvDisplayService.java#L284

So yes, I'm adding the SourceAndConverter object to the state directly. 

Thanks for linking the logic. I'll dig a bit more. ( and thanks for merging the PR )

**Comment by tpietzsch** [comment]
Yes, right, neither `SourceAndConverter` nor `Source` specify number of timepoints.

This only happens on the vistools layer (`BdvSource` has `getNumTimepoints()`).
The "problem" here is that you mix vistools and core sources. (BdvOverlaySource is vistools, SourceAndConverter is core).
Maybe an easy way around is to add the `SourceAndConverter` via vistools `BdvFunctions.show(yourSourceAndConverter, numTimepoints, Bdv.options().addTo(bdvh))` https://github.com/bigdataviewer/bigdataviewer-core/blob/783b7aee870348f231fdf8c7b214a5fa8a31e9df/src/main/java/bdv/util/BdvFunctions.java#L251-L255

Of course it would be better to make it more seamless to mix vistools and core. But I don't have a perfect solution at the moment.

**Comment by NicoKiaru** [comment]
Ah yes, thanks, I can try that. I think a minor inconvenience is that I will need to call this for each source in order to specify the timepoints instead of a single call with a List of sources. But maybe that's not a real problem.

**Comment by NicoKiaru** [comment]
Hi @tpietzsch just FYI, 

no need to change anything at the moment, but I wanted to share my reflexions after going back to this issue:

Using bdv vistools is way less performant than adding multiple sources. I did a bunch of tests and adding thousands of sources using bdv vistools takes a fraction of a sec per source, which is a lot. By contrast, adding a list of sourceandconverter using lower level APi is almost instantaneous.

Second, regarding overlays, that's a question of 'philosophy', but having a 'final' defined number of timepoint per overlay on creation is I think a mistake. Most of the time (at least that's what I do), you want the overlay to adapt to the current sources present in bdv. For instance you want to overlay the name of all sources - then it's the overlay which is in charge of adaptating its number of timepoints, and that can changed after being displayed.

**Comment by tpietzsch** [comment]
> Using bdv vistools is way less performant than adding multiple sources. I did a bunch of tests and adding thousands of sources using bdv vistools takes a fraction of a sec per source, which is a lot. By contrast, adding a list of sourceandconverter using lower level APi is almost instantaneous.

Could you make a small example that demonstrates this? I would like to fix this if possible...

**Comment by NicoKiaru** [comment]
I'm sorry, I don't have time to downsize my example now but here's how it looks with a demo that consists of showing an image where each pixel is a SourceAndConverter:

with the low level API https://github.com/NicoKiaru/ijp-imglib2bdvdemo-ij2/blob/58ec015908c15cb1e1e546277cd7a2591826fbf0/src/main/java/ch/epfl/biop/demos/DemoBigABBALogoCommand.java#L137 :

https://github.com/user-attachments/assets/2eee1659-82e6-4bf9-9912-bf2b326e5195

(takes about 30 secs)

with BdvFunctions.show https://github.com/NicoKiaru/ijp-imglib2bdvdemo-ij2/blob/101cc37d6b7b521804f3040dbdb969f761a7f975/src/main/java/ch/epfl/biop/demos/DemoBigABBALogoSlowCommand.java#L140

https://github.com/user-attachments/assets/11f09af4-7630-4d61-bed4-3c3f554b49c2

(takes about 1m30s) but actually the issue is that the time scales as N^2 there's probably a loop within a loop.

I have to admit that it's quite artificial - not everybody needs thousands of sources everyday, and the problem really shows up for nsources > 1 000, but there's not this problem by adding the sourceandconverters directly to the state, I'm pretty sure it scales linearly. 

Note: the demo is self contained - cloning the repo and running the demo works out of the box.

The logo consists of 70x70 sources (4900 sources)

**Comment by NicoKiaru** [comment]
The overlay workaround should be functional. Closing this.
